diff --git a/net/mptcp/protocol.h b/net/mptcp/protocol.h
index f799e51..4a2f84b 100644
--- a/net/mptcp/protocol.h
+++ b/net/mptcp/protocol.h
@@ -440,8 +440,6 @@ struct mptcp_sock {
 #endif
 };
 
-static_assert(sizeof(struct mptcp_sock) == sizeof(struct __orig_mptcp_sock));
-
 #define mptcp_data_lock(sk) spin_lock_bh(&(sk)->sk_lock.slock)
 #define mptcp_data_unlock(sk) spin_unlock_bh(&(sk)->sk_lock.slock)
 
